### PR TITLE
99 Add serializer for std::unique_ptr

### LINKED ISSUE #99: Implement serializers for std::unique_ptr<T>

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #99

### CODE REVIEW

**Comment by PhilMiller** [comment]
I could swear we had this in there already. Maybe in the wrong file or something?

**Comment by PhilMiller** [comment]
OK, I was thinking of the identical logic in `serializeRCP` I guess

File: src/checkpoint/container/unique_ptr_serialize.h [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+namespace checkpoint {
+
+template <typename Serializer, typename T>
+void serialize(Serializer& s, std::unique_ptr<T>& ptr) {
+  bool is_null = ptr == nullptr;
+  s | is_null;
+
+  if (not is_null) {
+    if (s.isUnpacking()) {
+      auto t = std::allocator<T>{}.allocate(1);
```
Comment (PhilMiller) [inline src/checkpoint/container/unique_ptr_serialize.h]: If we want to be *really* paranoid, we could add an assertion here that `alignof(T) <= __STDCPP_DEFAULT_NEW_ALIGNMENT__`

Comment (PhilMiller) [reply src/checkpoint/container/unique_ptr_serialize.h]: `static_assert` would be nice, but I'm not sure if that would run into SFINAE issues in this circumstance

**Review by PhilMiller — APPROVED** [review]

File: src/checkpoint/container/unique_ptr_serialize.h (hunk 1 above)
Comment (PhilMiller) [reply src/checkpoint/container/unique_ptr_serialize.h]: I think this allocation and reconstruction logic is likely duplicative of something that should be in functions called by `dispatch::deserialize` already. Let's make sure we get it right there, and then just call the same thing.

**Review by PhilMiller — CHANGES_REQUESTED** [review]

**Comment by lifflander** [comment]
@PhilMiller I've abstracted out allocation into `dispatch::Standard::allocate<T>`. I can't combine this with reconstruction easily. And we can't call into the API that returns a unique_ptr because that's only on the unpacking side.

**Pull request merged**